package-builder/flox-build.mk

```diff
@@ -0,0 +1,318 @@
+#
+# This makefile implements Tom's stepladder from manifest to Nix builds:
+#
+# 1. "local" (aka in-situ): sets $out in the environment, invokes the build
+#    commands in a subshell (using bash), then turns the $out directory into
+#    a Nix package with all outpath references replaced with the real $out
+#    and all bin/* commands wrapped with $FLOX_ENV/activate
+# 2. "sandbox": invokes that same script from within the runCommand builder,
+#    with no network and filesystem access and a fake home directory
+# 3. "sandbox with buildCache": does as above, with the build directory
+#    persisted across builds
+# 4. "staged": splits the builds into stages, each of which can be any of
+#    the above, and whose "locked" values are stored as a result symlink or
+#    as a storePath within the manifest
+#
+
+# Start by checking that the FLOX_ENV environment variable is set and that
+# we find the expected manifest.lock file in the FLOX_ENV directory.
+ifeq (,$(FLOX_ENV))
+  $(error FLOX_ENV not defined)
+endif
+MANIFEST_LOCK := $(FLOX_ENV)/manifest.lock
+ifeq (,$(wildcard $(MANIFEST_LOCK)))
+  $(error $(MANIFEST_LOCK) not found)
+endif
+
+# Substitute Nix store paths for packages required by this Makefile.
+__bashInteractive := @bashInteractive@
+__coreutils := @coreutils@
+__gitMinimal := @gitMinimal@
+__gnugrep := @gnugrep@
+__gnused := @gnused@
+__gnutar := @gnutar@
+__jq := @jq@
+__nix := @nix@
+
+# Access all required utilities by way of variables so that we don't depend
+# on anything from the user's PATH in the packaged version of flox. Note that
+# the __package_bin macro defined below will first test that the Nix package
+# substitution was successful, and if not then it will fall back to finding
+# the required tool from the PATH for use in the developer environment.
+__package_bin = $(if $(filter @%@,$(1)),$(2),$(1)/bin/$(2))
+_bash := $(call __package_bin,$(__bashInteractive),bash)
+_cp := $(call __package_bin,$(__coreutils),cp)
+_git := $(call __package_bin,$(__gitMinimal),git)
+_grep := $(call __package_bin,$(__gnugrep),grep)
+_jq := $(call __package_bin,$(__jq),jq)
+_nix := $(call __package_bin,$(__nix),nix)
+_readlink := $(call __package_bin,$(__coreutils),readlink)
+_rm := $(call __package_bin,$(__coreutils),rm)
+_sed := $(call __package_bin,$(__gnused),sed)
+_tar := $(call __package_bin,$(__gnutar),tar)
+_uname := $(call __package_bin,$(__coreutils),uname)
+
+# Identify path to build-manifest.nix, in same directory as this Makefile.
+_libexec_dir := $(realpath $(dir $(lastword $(MAKEFILE_LIST))))
```

Comment:
we call `realpath` in a couple places.
1. is this a "make realpath" or supposed to be the realpath bin ary
2. if its a binary, should it be `_coreutils/bin/realpath`?

Reply:
This is a gnumake builtin function: https://www.gnu.org/software/make/manual/make.html#index-realpath-1